apiserver/facades/controller/charmrevisionupdater/interface_test.go

```diff
@@ -0,0 +1,210 @@
+	return state
+}
+
+func makeResource(c *gc.C, name string, revision, size int, hexFingerprint string) resource.Resource {
+	fingerprint, err := resource.ParseFingerprint(hexFingerprint)
+	c.Assert(err, jc.ErrorIsNil)
+	return resource.Resource{
+		Meta: resource.Meta{
+			Name: name,
+			Type: resource.TypeFile,
+		},
+		Origin:      resource.OriginStore,
+		Revision:    revision,
+		Fingerprint: fingerprint,
+		Size:        int64(size),
+	}
+}
+
+// charmhubConfigMatcher matches only the charm IDs and revisions of a
+// charmhub.RefreshMany config.
+type charmhubConfigMatcher struct {
+	expected []charmhubConfigExpected
+}
+
+type charmhubConfigExpected struct {
+	id       string
+	revision int
+}
+
+func (m charmhubConfigMatcher) Matches(x interface{}) bool {
+	config, ok := x.(charmhub.RefreshConfig)
+	if !ok {
+		return false
+	}
+	request, _, err := config.Build()
+	if err != nil {
+		return false
+	}
+	if len(request.Context) != len(m.expected) || len(request.Actions) != len(m.expected) {
+		return false
+	}
+	for i, context := range request.Context {
+		if context.ID != m.expected[i].id {
+			return false
+		}
+		if context.Revision != m.expected[i].revision {
+			return false
+		}
+		action := request.Actions[i]
+		if *action.ID != m.expected[i].id {
+			return false
+		}
+	}
+	return true
+}
+
+func (charmhubConfigMatcher) String() string {
+	return "matches config"
+}
+
+func newFakeCharmstoreClient(st charmrevisionupdater.State) (charmstore.Client, error) {
+	charms := map[string]charmstoreCharm{
+		"mysql":     {name: "mysql", revision: 23},
+		"wordpress": {name: "wordpress", revision: 26},
+	}
+	client := &fakeCharmstoreClient{charms: charms}
+	return charmstore.NewCustomClient(client), nil
+}
+
+type charmstoreCharm struct {
+	name     string
+	revision int
+}
+
+type fakeCharmstoreClient struct {
+	charms map[string]charmstoreCharm
+}
+
+func (c *fakeCharmstoreClient) Latest(_ csparams.Channel, ids []*charm.URL, _ map[string][]string) ([]csparams.CharmRevision, error) {
+	revisions := make([]csparams.CharmRevision, len(ids))
+	for i, id := range ids {
+		charm, ok := c.charms[id.Name]
+		if !ok {
+			revisions[i] = csparams.CharmRevision{Err: errors.NotFoundf("charm %q", id.Name)}
+			continue
+		}
+		revisions[i] = csparams.CharmRevision{Revision: charm.revision}
+	}
+	return revisions, nil
+}
+
+func (c *fakeCharmstoreClient) ListResources(_ csparams.Channel, _ *charm.URL) ([]csparams.Resource, error) {
+	return nil, nil
+}
+
+func (c *fakeCharmstoreClient) GetResource(_ csparams.Channel, _ *charm.URL, _ string, revision int) (csclient.ResourceData, error) {
+	panic("not implemented")
+}
+
+func (c *fakeCharmstoreClient) ResourceMeta(_ csparams.Channel, _ *charm.URL, _ string, revision int) (csparams.Resource, error) {
+	panic("not implemented")
+}
+
+func (c *fakeCharmstoreClient) ServerURL() string {
+	panic("not implemented")
+}
+
+type facadeContextShim struct {
```

Comment:
I would do the following here:
- Pass `ctx.Auth()` to `NewCharmRevisionUpdaterAPIState`.
- Do `AuthController()` at the top of `UpdateLatestRevisions`.

This makes it more consistent with other APIs where the auth required might depend on the method called.

Then you don't need this mock; just the `authorizer`. `state` is never set for this in any case.